File: cpp/p67.cpp
#include "euler.h"

using namespace std;

int main() {
  ifstream file( "triangle.txt" );
  string line;
  vector< vector<int> > best_paths;

  while( getline( file, line ) ) {
    vector<string> cells = tokenize( line, " " );
    vector<int> row;
		
    for( int i = 0; i < cells.size(); i++ ) {
      int left_ancestor = 0, right_ancestor = 0;
			
      if( best_paths.size() > 0 && i > 0 )
	left_ancestor = best_paths.back()[ i - 1 ];
				
      if( best_paths.size() > 0 && best_paths.back().size() > i )
	right_ancestor = best_paths.back()[ i ];
				
      int best_path_thru_cell = max( left_ancestor, right_ancestor ) + string_to_number( cells[ i ] );
				
      row.push_back( best_path_thru_cell );
    }

    best_paths.push_back( row );
  }
	
  cout << *max_element( best_paths.back().begin(), best_paths.back().end() ) << endl;
}

File: clojure/p67.clj
(ns p67
  (:use [euler :only (max-of timed-test)])
  (:require [clojure.string :as str]))

(defn read-triangle []
  (for [line (str/split-lines (slurp "../data/triangle.txt"))]
    (for [token (str/split line #" ")]
      (Integer/parseInt token))))

(defn best-paths-row [row prev-best-paths-row]
  (for [i (range (count row))]
    (+ (nth row i) (max (nth prev-best-paths-row i 0)
			(nth prev-best-paths-row (dec i) 0)))))

(defn best-paths-triangle [[row & rows] triangle]
  (if-not row
    triangle
    (let [new-best-paths-row (best-paths-row row (peek triangle))]
      (recur rows (conj triangle new-best-paths-row)))))

(timed-test
 7273
 (max-of (peek (best-paths-triangle (read-triangle) []))))
